src/vos/tests/vts_aggregate.c

```diff
@@ -1017,6 +1017,200 @@ discard_13(void **state)
+			     &first_val);
+		if (first == AGG_PUNCH) {
+			/* Punch the first update */
+			do_punch(arg, type, oid, epr.epr_lo++, dkey, akey);
+		}
+	}
+
+	/** fake snapshot at epr.epr_lo, if first != AGG_NONE */
+	epoch = epr.epr_lo + 1;
+
+	for (i = 1; i <= NUM_INTERNAL; i++) {
+		bool	punch = (rand() % 5) == 0;
+
+		if (i == NUM_INTERNAL || punch) {
+			do_punch(arg, type, oid, epoch++, dkey, akey);
+			continue;
+		}
+		update_value(arg, oid, epoch++, dkey, akey,
+			     DAOS_IOD_SINGLE, sizeof(middle_val), NULL,
+			     &middle_val);
+		middle_epoch = epoch;
+	}
+
+	if (last == AGG_UPDATE) {
+		update_value(arg, oid, epoch++, dkey, akey, DAOS_IOD_SINGLE,
+			     sizeof(last_val), NULL, &last_val);
+	}
+
+	/* Set upper bound for aggregation */
+	epr.epr_hi = epoch++;
+
+	for (i = 0; i < 2; i++) {
+		if (discard)
+			rc = vos_discard(arg->ctx.tc_co_hdl, &epr);
+		else
+			rc = vos_aggregate(arg->ctx.tc_co_hdl, &epr);
+
+		assert_int_equal(rc, 0);
+
+		if (first != AGG_NONE) {
+			/* regardless of aggregate or discard, the first entry
+			 * should exist because it's outside of the epr.
+			 */
+			fetch_val = 0;
+			fetch_value(arg, oid, 1, dkey, akey, DAOS_IOD_SINGLE,
+				    sizeof(first_val), NULL, &fetch_val);
+			assert_int_equal(fetch_val, first_val);
+			/* Reading at "snapshot" should also work except for
+			 * punch, it will be gone.
+			 */
+			fetch_val = 0;
+			fetch_value(arg, oid, epr.epr_lo, dkey, akey,
+				    DAOS_IOD_SINGLE, sizeof(first_val), NULL,
+				    &fetch_val);
+			assert_int_equal(fetch_val,
+					 (first == AGG_PUNCH) ? 0 : first_val);
+		}
+
+		/* Intermediate value should be gone regardless but fetch will
+		 * get first_val if it's a discard
+		 */
+		expected = 0;
+		fetch_val = 0;
+		if (first == AGG_UPDATE && discard)
+			expected = first_val;
+		fetch_value(arg, oid, middle_epoch, dkey, akey, DAOS_IOD_SINGLE,
+			    sizeof(middle_val), NULL, &fetch_val);
+		assert_int_equal(fetch_val, expected);
+
+		/* Final value should be present for aggregation but not
+		 * discard
+		 */
+		fetch_val = 0;
+		fetch_value(arg, oid, epr.epr_hi, dkey, akey, DAOS_IOD_SINGLE,
+			    sizeof(last_val), NULL, &fetch_val);
+		expected = last_val;
+		if (discard) {
+			expected = 0;
+			if (first == AGG_UPDATE) {
+				/** Old value should still be there */
+				expected = first_val;
+			}
+		} else if (last != AGG_UPDATE) {
+			expected = 0;
+		}
+
+		assert_int_equal(fetch_val, expected);
+
+		/* One more test.  Punch the object at higher epoch, then
+		 * aggregate same epoch should get same results as the punch
+		 * is out of range.  Test is pointless for discard.
+		 */
+		if (discard)
+			break;
+
+		do_punch(arg, AGG_OBJ_TYPE, oid, epoch++, dkey, akey);
+	}
+
+	arg->ta_flags = old_flags;
+}
+
+/** Do a punch aggregation test */
+static void
+agg_punches_test(void **state, bool discard)
+{
+	int	first, last, type;
+	int	lstart;
+
+	daos_fail_loc_set(DAOS_VOS_AGG_RANDOM_YIELD | DAOS_FAIL_ALWAYS);
```

Comment:
The fail loc should be cleared after test done.

Reply:
done